### ISSUE TITLE
Need to know about the instance count dispalying while training

### ISSUE THREAD

**Issue by kevalp-weboccult** [ultralytics/ultralytics#20339]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello There 
I have one question for the instances displayed during model training, i am training a yolov8n-p6 segmentation model with 1920 images of size for some object segmentation, i am using batch size of 32, my every image contains aleast one instance, but when i check the Instances displaying for training epoch it shows 14, 17 some time 9 or 10 so what does it means and also if i am passing any args wrong or something 

Here is the training for the model, which shows the instances and ignores all the zeros since the object is very small so initially all values are zeros 

If you need, I can provide the args.yaml for model , also i am using latest ultralytics , my GPUS are very capable , like 24X2 48 GB gpu and CPU is also hight with 64 cores and 128 GB of ram and i am using latest pytorch and cuda , and also there are no dependancy issue 

![Image](https://github.com/user-attachments/assets/04df51e3-48cb-46e8-b93e-2fab3321e52b)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kevalp-weboccult, thank you for reaching out and using Ultralytics 🚀! This is an automated response to help get you started—an Ultralytics engineer will review your question and assist you soon.

We recommend checking the [Docs](https://docs.ultralytics.com/) for more information about training behavior in YOLO models, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions about training and instance counts are addressed there.

If you believe there may be a bug or unexpected behavior, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE)—for example, your `args.yaml`, a sample image, and any relevant training logs. This will help us diagnose the issue more effectively.

For best results with custom training, please include as much context as possible, like dataset samples and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth Q&A, check out [Discourse](https://community.ultralytics.com/)
- For sharing and discussion, explore our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Ensure you are on the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your detailed question and screenshot! An Ultralytics engineer will follow up soon to help clarify the instance count behavior during training.

**Comment by glenn-jocher** [comment]
Hi @kevalp-weboccult,

The "Instances" value in the training logs represents the average number of objects per batch that are actively contributing to the loss calculation after preprocessing and filtering operations. 

Since you're using a batch size of 32, this number is showing the average objects being used for training across those 32 images in each batch. It doesn't represent the total number of instances in your dataset. Some reasons for the lower numbers:

1. Very small objects might be filtered out if they're below certain size thresholds
2. Objects that don't meet anchor criteria during training
3. Objects may be dropped during augmentation 
4. The model might only process a subset of objects per image for efficiency

This number fluctuates between batches and is normal behavior - it doesn't indicate any problem with your training setup. The model is still learning from the valid instances it processes.

**Comment by Y-T-G** [comment]
It's the number of instances in a batch which is randomly generated by combining multiple images.

**Comment by kevalp-weboccult** [comment]
@glenn-jocher  i have removed all the augmentation and filters which eliminates the small objects due to area_ration and widht height ratio, so there are no chances for the 1 and 2 nd point can you please explain the 4th point in brief and for the 2nd point what should i do for that like i am using p6 head for small object segmentation and image size of 1920 for 2560 image size

**Comment by glenn-jocher** [comment]
Hi @kevalp-weboccult,

Regarding point #4, during training, YOLO models process a limited number of objects per batch due to internal memory optimization. The model selects a subset of the highest-confidence detections per image to compute loss, rather than using every single instance. This selection happens dynamically during the forward pass.

For the anchor criteria issue: P6 models (with additional detection layers) are indeed better for small objects, but you can still experience filter-out effects during training. The model dynamically assigns objects to anchors during training based on shape compatibility. Some very small objects might not match well with the predefined anchors, even with P6.

You could try:
1. Using `model.args.box *= 3` to scale box loss and increase importance of small objects
2. Adding `rect=False` if you're using rectangular training 
3. Consider adding `multi_scale=True` which can help with small object detection by varying image sizes during training

These should help increase the number of instances being processed during training.
